### PR TITLE
New Xeno Caste: Defiler

### PR DESCRIPTION
**Pull request by Surrealaser**
:cl: Surrealistik, Sprites by Manezinho
add: New T3 evolution added for Hivelord and Carrier: The Defiler. Has above average to high health, speed, melee damage and armor. Can vent crawl.
/:cl:
Has three abilities: Emit Neurogas, Toggle Neurospines and Defile.
Emit Neurogas discharges 3 small neurotoxin clouds centered on the Defiler after a 2 second channel with a 1 second delay between each discharge. Costs 200 plasma and has a 40 second cooldown. 
Toggle Neurospines toggles on/off injecting 5 neurotoxin on throttle/slash at the cost of 50 plasma per attack. 
Defile injects 15 neurotoxin and larval growth toxin + 2 more 10u doses in 1.5 second channeled intervals after. On the third interval, it implants a larva. Costs 150 Plasma and has a 20 second cooldown.

### CODE REVIEW

**Comment by drawsstuff** [comment]
[Nature at its finest](https://i.imgur.com/MP8y04Z.jpg)

File: code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+
+	// *** Defense *** //
+	armor_deflection = 47
+
+	// *** Defiler Abilities *** //
+
+/datum/xeno_caste/defiler/ancient
+	upgrade_name = "Ancient"
+	caste_desc = "Being within mere eyeshot of this hulking, dripping monstrosity fills you with a deep, unshakeable sense of unease."
+	ancient_message = "You are the ultimate alien rapist. You will impregnate the marines, see them burst open before you, and hear the gleeful screes of your larvae."
```
Comment (factoryman942) [inline code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm]: i am uncomfortable

Comment (Ghommie) [reply code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm]: I did miss this part but, dear revioli god, tone the acient message flavour down and use euphisms, just a little, or I'm giving marines pepper sprays.

Comment (Surrealaser) [reply code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm]: Literally no one sees it but the player; calm your tits

Comment (Ghommie) [reply code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm]: Understood, after all you aren't wrong.

Comment (topkek1) [reply code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm]: Ghommie, chill out, please jesus christ

Comment (Ghommie) [reply code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm]: Already done, apologises.

Comment (factoryman942) [reply code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm]: even if only the defiler player sees it, someone's going to end up with a rule 8 at some point because of this, i guarantee it

**Comment by DominikPanic** [comment]
Conflicts. I would also recommend toning down that ancient message.

File: code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+	pixel_x = -16
+	old_x = -16
+	wound_type = "runner" //used to match appropriate wound overlays
+	tier = 3
+	upgrade = 0
+	var/datum/effect_system/smoke_spread/smoke_system = null
+	var/last_defiler_sting = null
+	var/last_emit_neurogas = null
+	var/last_use_neuroclaws = null
+	var/neuro_claws = FALSE
```
[the remark this answers is not in the document]
Comment (Surrealaser) [reply code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm]: Nah, better to keep it local.

**Comment by Alterist** [comment]
sprites are nice, please delete the defiler.dmi if its not being used

File: code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm [hunk 3]
```diff
@@ -0,0 +1,198 @@
+/datum/xeno_caste/defiler
+	caste_name = "Defiler"
+	display_name = "Defiler"
+	upgrade_name = "Young"
+	caste_desc = "A frightening looking, bulky xeno that drips with suspect green fluids."
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm]:
```suggestion
	caste_desc = "A frightening looking, bulky xeno that drips with suspicious green fluids."
```

File: code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+	update_fire() //the fire overlay depends on the xeno's stance, so we must update it.
+	update_wounds()
+
+/mob/living/carbon/Xenomorph/Defiler/throw_item(atom/target)
+	throw_mode_off()
+
+/mob/living/carbon/Xenomorph/Defiler/hitby(atom/movable/AM as mob|obj,var/speed = 5)
+	if(ishuman(AM))
+		return
+	..()
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm]:
```suggestion
	return ..()
```

File: code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+/mob/living/carbon/Xenomorph/Defiler/Bumped(atom/movable/AM as mob|obj)
+	if(emitting_gas) //We don't get bumped around
+		return
+	..()
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm]:
```suggestion
	return ..()
```

File: code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+/mob/living/carbon/Xenomorph/Defiler/throw_item(atom/target)
+	throw_mode_off()
+
+/mob/living/carbon/Xenomorph/Defiler/hitby(atom/movable/AM as mob|obj,var/speed = 5)
+	if(ishuman(AM))
+		return
+	..()
+
+/mob/living/carbon/Xenomorph/Defiler/Bumped(atom/movable/AM as mob|obj)
+	if(emitting_gas) //We don't get bumped around
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm]: Even by tanks?

File: code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm (hunk 3 above)
Comment (Surrealaser) [reply code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm]: No, I like suspect more:

1.
not to be relied on or trusted; possibly dangerous or false.

File: code/modules/mob/living/carbon/xenomorph/Powers.dm [hunk 7]
```diff
@@ -2073,4 +2079,149 @@
...
-			A.update_button_icon()
\ No newline at end of file
+			A.update_button_icon()
+
+
+
+//Defiler abilities
+/mob/living/carbon/Xenomorph/Defiler/proc/emit_neurogas()
+
+	if(!check_state()) return
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/Powers.dm]: Same line control statement.

File: code/modules/mob/living/carbon/xenomorph/Powers.dm [hunk 8]
```diff
@@ -2073,4 +2079,149 @@
...
+		return
+	playsound(loc, 'sound/effects/smoke.ogg', 25)
+	var/turf/T = get_turf(src)
+	smoke_system = new /datum/effect_system/smoke_spread/xeno_weaken()
+	smoke_system.amount = 2
+	smoke_system.set_up(2, 0, T)
+	smoke_system.start()
+	T.visible_message("<span class='danger'>Noxious smoke billows from the hulking xenomorph!</span>")
+	spawn(10)
+		dispense_gas(count + 1)
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/Powers.dm]: Do this iteratively, not recursively.

File: code/modules/mob/living/carbon/xenomorph/Powers.dm [hunk 9]
```diff
@@ -2073,4 +2079,149 @@
...
+	face_atom(H)
+	if(!do_after(src, DEFILER_STING_CHANNEL_TIME, TRUE, 5, BUSY_ICON_HOSTILE))
+		return
+	animation_attack_on(H)
+	playsound(H, pick('sound/voice/alien_drool1.ogg', 'sound/voice/alien_drool2.ogg'), 15, 1)
+	H.reagents.add_reagent("xeno_toxin", DEFILER_STING_AMOUNT_RECURRING) //10 units transferred.
+	H.reagents.add_reagent("xeno_growthtoxin", DEFILER_STING_AMOUNT_RECURRING)
+
+	if(count < 2)
+		defiler_recurring_injection(H, count + 1)
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/Powers.dm]: Iteration, not recursion.

File: code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm (hunk 3 above)
Comment (spookydonut) [reply code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm]: suspect is correct here

File: code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm (hunk 2 above)
[the remark this answers is not in the document]
Comment (spookydonut) [reply code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm]: these are fine.

File: code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm (hunk 6 above)
Comment (spookydonut) [reply code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm]: Isn't there a tank_act or something

File: code/modules/mob/living/carbon/xenomorph/Xenomorph.dm [hunk 10]
```diff
@@ -111,6 +111,11 @@
 	health = maxHealth
 	speed = xeno_caste.speed
 
+/mob/living/carbon/Xenomorph/Defiler/set_datum()
+	. = ..()
+	var/datum/xeno_caste/defiler/neuro_upgrade = xeno_caste_datums[caste_base_type][CLAMP(upgrade + 1, 1, 4)]
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/Xenomorph.dm]: uh what is this?

File: code/modules/reagents/chemistry_reagents/other.dm [hunk 11]
```diff
@@ -817,3 +817,19 @@
 
 /datum/reagent/xeno_neurotoxin/overdose_crit_process(mob/living/M)
 		M.adjustOxyLoss(min(4,volume * 0.2 * REM)) //Overdose starts applying more oxy damage
+
+/datum/reagent/xeno_growthtoxin
```
Comment (spookydonut) [inline code/modules/reagents/chemistry_reagents/other.dm]: shouldnt this be in the other PR not this one?

File: code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm (hunk 6 above)
Comment (Rohesie) [reply code/modules/mob/living/carbon/xenomorph/Castes/Defiler.dm]: Ah, the tank handles it through `/obj/vehicle/multitile/hitbox/cm_armored/Bump(var/atom/A)`, but it doesn't really call for `Bumped()` so this won't be an issue.

File: code/modules/mob/living/carbon/xenomorph/Castes/Sentinel.dm [hunk 12]
```diff
@@ -226,6 +226,8 @@
 
 
 /mob/living/carbon/Xenomorph/Sentinel/proc/neurotoxin_recurring_injection(var/mob/living/H)
+	if(!Adjacent(H) || stagger)
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/Castes/Sentinel.dm]: This kind of bug fix should be in a different PR.

File: code/modules/mob/living/carbon/xenomorph/Powers.dm [hunk 13]
```diff
@@ -2077,4 +2083,149 @@
...
+		return
+
+	round_statistics.defiler_neurogas_uses++
+
+	visible_message("<span class='xenodanger'>[src] emits a noxious gas!</span>", \
+	"<span class='xenodanger'>You emit neurogas!</span>")
+	dispense_gas(3)
+
+/mob/living/carbon/Xenomorph/Defiler/proc/dispense_gas(count = 0)
+	while(count)
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/Powers.dm]: Make sure to `set waitfor = FALSE` if you don't want to this to lock the other procs, but make sure the caller does not require a return from the proc.
`count = 0` does no longer seem necessary as a passed argument, now does it?

File: code/modules/mob/living/carbon/xenomorph/Xenomorph.dm (hunk 10 above)
Comment (Surrealaser) [reply code/modules/mob/living/carbon/xenomorph/Xenomorph.dm]: Neuroclaw injections now scale

File: code/modules/reagents/chemistry_reagents/other.dm (hunk 11 above)
Comment (Surrealaser) [reply code/modules/reagents/chemistry_reagents/other.dm]: No, because it's both integral and exclusive to the Defiler

File: code/modules/mob/living/carbon/xenomorph/Powers.dm (hunk 13 above)
Comment (Rohesie) [reply code/modules/mob/living/carbon/xenomorph/Powers.dm]: It might be github, but is this properly tabbed?

File: code/modules/mob/living/carbon/xenomorph/Powers.dm (hunk 7 above)
Comment (Rohesie) [reply code/modules/mob/living/carbon/xenomorph/Powers.dm]: This is still a thing.

**Comment by plapatin** [comment]
brap

File: code/game/objects/items/storage/pouch.dm [hunk 14]
```diff
@@ -93,23 +93,25 @@
...
 					"/obj/item/reagent_container/hypospray/autoinjector/tramadol",
 					"/obj/item/reagent_container/hypospray/autoinjector/tricordrazine",
+					"/obj/item/reagent_container/hypospray/autoinjector/hypervene",
 					"/obj/item/stack/medical/bruise_pack",
 					"/obj/item/stack/medical/splint"
 					)
 
 /obj/item/storage/pouch/firstaid/full
-	desc = "Contains a painkiller autoinjector, first-aid autoinjector, some ointment, and some bandages."
+	desc = "Contains a painkiller autoinjector, first-aid autoinjector, hypervene autoinjector, splints, some ointment, and some bandages."
```
Comment (Rohesie) [inline code/game/objects/items/storage/pouch.dm]: This is out of scope.
Again, find a way to introduce your caste without its balance requiring the addition of hypervene to the FAP.

Comment (Surrealaser) [reply code/game/objects/items/storage/pouch.dm]: Um first of all, it doesn't necessarily require this, it's 

A: an experiment.

and

B: it makes sense. Marines should have countermeasures to known deadly toxins on their person due to the risk involved and their time critical nature.

Comment (Surrealaser) [reply code/game/objects/items/storage/pouch.dm]: This has been 'fixed'. The growth bonus of growth toxin has been reduced and the autoinjectors have been removed.

**Comment by Rohesie** [comment]
* Rudimentary transformation doesn't include Defilers.

File: code/modules/mob/living/carbon/xenomorph/Embryo.dm [hunk 15]
```diff
@@ -81,6 +83,10 @@
 		counter += 1 + max(0,(0.03 * affected_mob.health)) //Up to +300% faster, depending on the health of the host
 	else if(stage <= 4)
 		counter++
+	if(isliving(affected_mob) && !stasis) //Cold temperatures and stasis prevent the growth toxin from having an effect.
+		var/mob/living/L = affected_mob
+		if(L.reagents.get_reagent_amount("xeno_growthtoxin"))
+			counter += 5 //Dramatically accelerates larval growth. You don't want this stuff in your body. Larva hits Stage 5 in just over 3 minutes, assuming the victim has growth toxin for the full duration.
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/Embryo.dm]:
```suggestion
			counter += 4 //Dramatically accelerates larval growth. You don't want this stuff in your body. Larva hits Stage 5 in 4 minutes, assuming the victim has growth toxin for the full duration.
```

**Comment by Rohesie** [comment]
Waiting on `simplemake` changes as well.

**Comment by Moja10** [comment]
Excuse my stupidity but is this a caste capable of creating larva that isn't a Queen? As in, it is capable of creating larva within a human without involving the Queen's eggs at all. Seems like a slippery slope to me, like if the carrier could just conjure huggers out of her ass.

**Comment by Alterist** [comment]
The carrier can conjure huggers out of her ass

**Comment by Moja10** [comment]
muh slippery slope
